hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -1674,13 +1907,38 @@ def draw_boolean(
             assert p < (1 - 2 ** (-64))
 
         kwargs: BooleanKWargs = {"p": p}
+
+        if self.ir_tree_leaves is not None and observe:
+            forced = self._pop_ir_tree_value("boolean", kwargs)
+
         value = self.provider.draw_boolean(**kwargs, forced=forced)
         if observe:
             self.observer.draw_boolean(
                 value, kwargs=kwargs, was_forced=forced is not None
             )
+            self.ir_tree.draw_boolean(value, kwargs, was_forced=forced is not None)
         return value
 
+    def _pop_ir_tree_value(self, ir_type: IRTypeName, kwargs: IRKWargsType) -> IRType:
+        assert self.ir_tree_leaves is not None
+
+        if self.ir_tree_leaves_index >= len(self.ir_tree_leaves):
+            self.mark_overrun()
+
+        leaf = self.ir_tree_leaves[self.ir_tree_leaves_index]
+        if leaf.ir_type != ir_type or kwargs != leaf.kwargs:
+            # Unlike buffers, not every ir tree is a valid choice sequence. If
+            # we want to maintain determinism we don't have many options here
+            # beyond giving up when a modified tree becomes misaligned.
+            #
+            # For what it's worth, misaligned buffers — albeit valid — are rather
+            # unlikely to be *useful* buffers, so this isn't an enormous
+            # downgrade.
+            self.mark_overrun()
```

Comment:
shrinking may make modifications that misalign the ir tree. When working with buffers, we still run these all the way to a result, but my guess is they are very unlikely to be useful shrinks.

I don't know if there's anything better we can do for ir trees than giving up, beyond realigning tricks that are research papers in their own right. One idea is passing `shrinker.engine.random` here and continuing randomly, which might work, but I don't know if the same ir tree prefix running to different buffers will blow up. 

By bailing out early we do less work in the test function, which might mean we actually spend more time proportionally on useful shrinks by overrunning. So it might be a wash at the end of the day. But I don't have a good intuition.